### ISSUE TITLE
Logging SQL queries

### ISSUE THREAD

**Issue by fortm** [begriffs/postgrest#735]
My current understanding is that postgrest creates optimal SQL queries and can use Stored procedures for complex ER diagram usecases. For external dependencies , we can use NOTIFY and LISTEN. 

Can we log SQL queries that are called from URL like these - 
http://localhost:3000/competition?select=name,film{title}

I wanted to verify if postgrest will do ( select * ) or only ( select columns that are provided explicitly ) 

In general , also wanted to have ability to check SQL queries being called from Stored procedure in development or is there better debugging alternative for Stored procedures inside Postgres ?

**Comment by ruslantalpa** [comment]
Logging queries is the job of the db, and you can enable that if you want in postgresql.
Postgrest will select only the specified colums.
You will not see the internal queries from the storeed procedures because its assumed you wrote those procs and know their internals (when you see a call to one being made in the db log)
Do not "debug" storeed procedures by calling them over postgrest, use a normal sql client and work with them there

**Comment by fortm** [comment]
@ruslantalpa, i have another question, I am not sure if I should create another issue for that.  But it is about caching the results from postgrest in redis server. 

I am looking to use postgrest behind another service that faces client. And am looking for caching strategy for results that we get from db. Possibly by storing key as postgrest URL and value as resulting JSON inside Redis. This can be invalidated from external API based on NOTIFY raised from postgres Stored procedures.

Do you think this is a good way to cache  SQL results given by postgrest ?

**Comment by ruslantalpa** [comment]
If another service is facing the clients why does not that service run queries directly, why go through postgrest and not run the queries directly against the db?
Postgrest is meant to face the clients (but  behind a proxy).
If you are just using it as an easy way generate a few fixed queries and you are not exposing it's full power to the clients you are better of writing those queries manually and running them agains the db. 

The cache key is application specific, maybe even request specific so i can't comment on how that key should be calculated. Invalidation however can be done without the listen/notify.
Whenever you have a post/patch request entering the system you know it's going to change something so you should invalidate the caches  at that stage (in the proxy, before the request reaches postgrest) and not implement triggers for every table just to invalidate caches.

**Comment by ruslantalpa** [comment]
an example of a function in the proxy calculating a cache key https://github.com/ruslantalpa/sub0_sample_app/blob/master/user_module.lua#L8-L30

**Comment by SebAlbert** [comment]
I had the same thought on "why using PostgREST then". However, I think your suggestion for cache invalidation is only feasible if the service (client-facing) is the only thing that manipulates data, which might not be the case, so a LISTEN/NOTIFY approach seems sensible to me.

**Comment by ruslantalpa** [comment]
Considering that you need the flexibility of postgrest then i would have everyone accessing the db through postgrest (with proxy in front), each client with it's own rights and even api's.
and the proxy would be responsible for cache for everyone.
The proxy could also implement a small api for invalidating cache and the (few) things that modify the data directly in the db should be smart enough to know what they are changing and call that api to invalidate the caches. Having listen/notify on every table will be harder plus knowing what data changed (rows) does not always give you the entire info necessary to know what caches to invalidate (like the http request).
The way to implement cache is highly dependent on the system so sometimes the above does not apply, sometimes this works better https://vimeo.com/145847021 and i think it's better then using notify

**Comment by fortm** [comment]
Thanks @ruslantalpa, @SebAlbert 
The vimeo link seems like uses kafka queue for cache invalidation and postgres log detection instead of triggers / notify. I think this will stream every insert / update / delete as they happen and it is upto application listening to queue to react to each of these event or skip some of them.

I have graphql schema in server side behind which I am using postgrest. So I am not behind a full fledged client facing service, only a thin validation layer.
graphql schema has now batching logic for REST API using data-loader and therefore making REST calls from schema is batched and made only once provided same URL is hit twice in any nested query. For SQL calls, graphql still can lead to too many SQL calls for even single nested request

**Comment by ruslantalpa** [comment]
How manny rest calls (to postgrest) will your resolvers make for something like this?
```
{
	project(id:1){
		id
		name
		client {
			id
			name
		}
		tasks {
			id
			name
		}
	}
}

```

**Comment by fortm** [comment]
@ruslantalpa ,  assuming here we have a project, client , project_client and project_task tables.
Project can have many client and many tasks. 

With 3 resolvers for ProjectType, List<ClientType>, and List<TaskType >:
```
ProjectType resolves to : http://localhost:3000/project?select(id,name)&id=eq.1
ClientType resolves to :   http://localhost:3000/project_client?select=client{id,name}&project_id=eq.1
Each TaskType resolves to :    http://localhost:3000/project_task?select{id,name}&project_id=eq.1
```
Here for a GRAPHQL request like above, there will be 3 REST Calls, one for resolving each of nodes i.e a project's id and name, followed by its client's id and name and lastly for all tasks that belong to the project id  = 1

I think with slight changes in GRAPHQL request, we will start to see issue with batching as in - 

```
{
    projects {
        id
        name
        client {
            id
            name
        }
        tasks {
            id
            name
        }
    }
}
```

In these cases, since we are iterating over summaries of projects whose id would be 1, 2, 3..
instead of making REST calls in this way - 

```
http://localhost:3000/project_client?select=client{id,name}&project_id=eq.1
http://localhost:3000/project_client?select=client{id,name}&project_id=eq.2
```

it can be batched using dataloader in resolver for client to :  

`http://localhost:3000/project_client?select=client{id,name}&project_id=in.1,2`

I also agree that postgrest is being under utilised here since I could have done just  1 REST call to get these directly from postgREST - 

`localhost:3000/project?select=id,name,client{id,name},project_task{id,name}&id=eq.1`

**Comment by ruslantalpa** [comment]
Postgrest indeed can be used to fetch everything in one go, this is why the `&select` was introduced, to be used as a base for graphql.
The example i gave is a simple one, imagine you want to fetch a list of projects with the client for each one and the tasks for each one, and you have about 100 projects, now count your rest requests if you do it like you explained. Dataloader won't save you.
You have to implement resolvers that fetch the data for the entire tree in one go, not just the current type, i.e you have to analyse the entire AST in your resolvers.
And i guess you've been paying attention to the link i posted earlier :)
https://github.com/ruslantalpa/sub0_sample_app/

**Comment by begriffs** [comment]
@fortm I added docs for you on how to get detailed SQL logs  - http://postgrest.com/en/v0.4/admin.html#debugging

Closing this issue as the original concern is addressed. Feel free to open more issues of course for separate concerns.
